File: entity-store/src/main/java/jetbrains/exodus/entitystore/iterate/PropertyValueIterable.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.entitystore.iterate;

import jetbrains.exodus.ByteIterable;
import jetbrains.exodus.bindings.ComparableBinding;
import jetbrains.exodus.bindings.ComparableSet;
import jetbrains.exodus.bindings.LongBinding;
import jetbrains.exodus.entitystore.*;
import jetbrains.exodus.entitystore.tables.PropertyTypes;
import jetbrains.exodus.env.Cursor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Iterates all entities of specified entity type having specified property equal to a value.
 */
@SuppressWarnings({"unchecked"})
public final class PropertyValueIterable extends PropertyRangeOrValueIterableBase {

    @NotNull
    private final Comparable value;

    static {
        registerType(getType(), new EntityIterableInstantiator() {
            @Override
            public EntityIterableBase instantiate(PersistentStoreTransaction txn, PersistentEntityStoreImpl store, Object[] parameters) {
                try {
                    return new PropertyValueIterable(txn,
                        Integer.valueOf((String) parameters[0]), Integer.valueOf((String) parameters[1]),
                        Long.parseLong((String) parameters[2]));
                } catch (NumberFormatException e) {
                    return new PropertyValueIterable(txn,
                        Integer.valueOf((String) parameters[0]), Integer.valueOf((String) parameters[1]),
                        (Comparable) parameters[2]);
                }
            }
        });
    }

    public PropertyValueIterable(@NotNull final PersistentStoreTransaction txn,
                                 final int entityTypeId,
                                 final int propertyId,
                                 @NotNull final Comparable value) {
        super(txn, entityTypeId, propertyId);
        this.value = PropertyTypes.toLowerCase(value);
    }

    @Override
    @NotNull
    public EntityIteratorBase getIteratorImpl(@NotNull final PersistentStoreTransaction txn) {
        final EntityIterableBase it = getPropertyValueIndex();
        if (it.isCachedInstance()) {
            final UpdatablePropertiesCachedInstanceIterable cached = (UpdatablePropertiesCachedInstanceIterable) it;
            if (value.getClass() != cached.getPropertyValueClass()) {
                return EntityIteratorBase.EMPTY;
            }
            return cached.getPropertyValueIterator(value);
        }
        final Cursor valueIdx = openCursor(txn);
        if (valueIdx == null) {
            return EntityIteratorBase.EMPTY;
        }
        return new PropertyValueIterator(valueIdx);
    }

    @Override
    public boolean nonCachedHasFastCountAndIsEmpty() {
        return true;
    }

    @Override
    @NotNull
    protected EntityIterableHandle getHandleImpl() {
        final int entityTypeId = getEntityTypeId();
        final int propertyId = getPropertyId();
        return new ConstantEntityIterableHandle(getStore(), PropertyValueIterable.getType()) {

            @NotNull
            @Override
            public int[] getPropertyIds() {
                return new int[]{propertyId};
            }

            @Override
            public void toString(@NotNull final StringBuilder builder) {
                super.toString(builder);
                builder.append(entityTypeId);
                builder.append('-');
                builder.append(propertyId);
                builder.append('-');
                builder.append(value.toString());
            }

            @Override
            public void hashCode(@NotNull final EntityIterableHandleHash hash) {
                hash.apply(entityTypeId);
                hash.applyDelimiter();
                hash.apply(propertyId);
                hash.applyDelimiter();
                hash.apply(value.toString());
            }

            @Override
            public int getEntityTypeId() {
                return PropertyValueIterable.this.getEntityTypeId();
            }

            @Override
            public boolean isMatchedPropertyChanged(@NotNull final EntityId id,
                                                    final int propId,
                                                    @Nullable final Comparable oldValue,
                                                    @Nullable final Comparable newValue) {
                //noinspection OverlyComplexBooleanExpression
                return propertyId == propId && entityTypeId == id.getTypeId() && (isValueMatched(oldValue) || isValueMatched(newValue));
            }

            private boolean isValueMatched(Comparable value) {
                if (value == null) {
                    return false;
                }
                if (value instanceof ComparableSet) {
                    return ((ComparableSet) value).containsItem(PropertyValueIterable.this.value);
                }
                value = PropertyTypes.toLowerCase(value);
                return value.compareTo(PropertyValueIterable.this.value) == 0;
            }
        };
    }

    @Override
    protected long countImpl(@NotNull final PersistentStoreTransaction txn) {
        final ByteIterable key = getStore().getPropertyTypes().dataToPropertyValue(value).dataToEntry();
        final Cursor valueIdx = openCursor(txn);
        return valueIdx == null ? 0 : new SingleKeyCursorCounter(valueIdx, key).getCount();
    }

    @Override
    public boolean isEmptyImpl(@NotNull PersistentStoreTransaction txn) {
        final ByteIterable key = getStore().getPropertyTypes().dataToPropertyValue(value).dataToEntry();
        final Cursor valueIdx = openCursor(txn);
        return valueIdx == null || new SingleKeyCursorIsEmptyChecker(valueIdx, key).isEmpty();
    }

    private static EntityIterableType getType() {
        return EntityIterableType.ENTITIES_BY_PROP_VALUE;
    }

    private final class PropertyValueIterator extends EntityIteratorBase {

        private boolean hasNext;
        @NotNull
        private final ComparableBinding binding;

        private PropertyValueIterator(@NotNull final Cursor cursor) {
            super(PropertyValueIterable.this);
            setCursor(cursor);
            binding = getStore().getPropertyTypes().dataToPropertyValue(value).getBinding();
            final ByteIterable key = binding.objectToEntry(value);
            checkHasNext(getCursor().getSearchKey(key) != null);
        }

        @Override
        public boolean hasNextImpl() {
            return hasNext;
        }

        @Override
        @Nullable
        public EntityId nextIdImpl() {
            if (hasNextImpl()) {
                explain(getType());
                final Cursor cursor = getCursor();
                final EntityId result = new PersistentEntityId(getEntityTypeId(), LongBinding.compressedEntryToLong(cursor.getValue()));
                checkHasNext(cursor.getNextDup());
                return result;
            }
            return null;
        }

        private void checkHasNext(final boolean success) {
            hasNext = success && value.compareTo(binding.entryToObject(getCursor().getKey())) == 0;
        }
    }
}

File: persistent-entity-store/src/main/kotlin/jetbrains/exodus/entitystore/iterate/PropertyValueIterable.kt
/*
 * Copyright ${inceptionYear} - ${year} ${owner}
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.entitystore.iterate

import jetbrains.exodus.ArrayByteIterable
import jetbrains.exodus.bindings.ComparableSet
import jetbrains.exodus.bindings.LongBinding
import jetbrains.exodus.entitystore.*
import jetbrains.exodus.entitystore.tables.PropertyTypes
import jetbrains.exodus.env.Cursor

class PropertyValueIterable(
    txn: StoreTransaction,
    entityTypeId: Int,
    propertyId: Int,
    value: Comparable<*>
) : PropertyRangeOrValueIterableBase(txn, entityTypeId, propertyId) {

    private val value = PropertyTypes.toLowerCase(value)
    private val valueClass = value.javaClass
    private var forceCached = false

    companion object {
        init {
            registerType(EntityIterableType.ENTITIES_BY_PROP_VALUE) { txn, _, parameters ->
                try {
                    return@registerType PropertyValueIterable(
                        txn,
                        (parameters[0] as String).toInt(),
                        (parameters[1] as String).toInt(),
                        (parameters[2] as String).toLong()
                    )
                } catch (e: NumberFormatException) {
                    return@registerType PropertyValueIterable(
                        txn,
                        (parameters[0] as String).toInt(),
                        (parameters[1] as String).toInt(),
                        parameters[2] as Comparable<*>
                    )
                }
            }
        }
    }

    override fun canBeCached() = forceCached || value is Boolean || super.canBeCached()

    override fun getIteratorImpl(txn: StoreTransaction): EntityIteratorBase {
        val it = propertyValueIndex
        if (it.isCachedInstance) {
            val cached = it as UpdatablePropertiesCachedInstanceIterable
            return if (value.javaClass != cached.getPropertyValueClass()) {
                EntityIteratorBase.EMPTY
            } else {
                cached.getPropertyValueIterator(value as Comparable<Any>)
            }
        }
        val valueIdx = openCursor(txn) ?: return EntityIteratorBase.EMPTY
        return PropertyValueIterator(valueIdx, reverse = false)
    }

    override fun getReverseIteratorImpl(txn: StoreTransaction): EntityIterator {
        val it = propertyValueIndex
        if (it.isCachedInstance) {
            val cached = it as UpdatablePropertiesCachedInstanceIterable
            return if (value.javaClass != cached.getPropertyValueClass()) {
                EntityIteratorBase.EMPTY
            } else {
                forceCached = true
                getOrCreateCachedInstance(txn).getReverseIteratorImpl(txn)
            }
        }
        val valueIdx = openCursor(txn) ?: return EntityIteratorBase.EMPTY
        return PropertyValueIterator(valueIdx, reverse = true)
    }

    override fun nonCachedHasFastCountAndIsEmpty() = true

    override fun getHandleImpl(): EntityIterableHandle {
        val entityTypeId = entityTypeId
        val propertyId = propertyId
        return object : ConstantEntityIterableHandle(store, EntityIterableType.ENTITIES_BY_PROP_VALUE) {

            override fun getPropertyIds() = intArrayOf(propertyId)

            override fun toString(builder: StringBuilder) {
                super.toString(builder)
                builder.append(entityTypeId)
                builder.append('-')
                builder.append(propertyId)
                builder.append('-')
                builder.append(value.toString())
            }

            override fun hashCode(hash: EntityIterableHandleHash) {
                hash.apply(entityTypeId)
                hash.applyDelimiter()
                hash.apply(propertyId)
                hash.applyDelimiter()
                hash.apply(value.toString())
            }

            override fun getEntityTypeId() = entityTypeId

            override fun isMatchedPropertyChanged(
                id: EntityId,
                propId: Int,
                oldValue: Comparable<*>?,
                newValue: Comparable<*>?
            ): Boolean {
                return propertyId == propId && entityTypeId == id.typeId && (isValueMatched(oldValue) || isValueMatched(
                    newValue
                ))
            }

            private fun isValueMatched(value: Comparable<*>?): Boolean {
                value ?: return false
                if (value is ComparableSet<*>) {
                    return value.containsItem(this@PropertyValueIterable.value)
                }
                if (value.javaClass != valueClass) return false
                return PropertyTypes.toLowerCase(value).compareTo(this@PropertyValueIterable.value) == 0
            }
        }
    }

    override fun countImpl(txn: StoreTransaction): Long {
        val key = storeImpl.propertyTypes.dataToPropertyValue(value).dataToEntry()
        val valueIdx = openCursor(txn)
        return if (valueIdx == null) 0 else SingleKeyCursorCounter(valueIdx, key).count
    }

    override fun isEmptyImpl(txn: StoreTransaction): Boolean {
        val key = storeImpl.propertyTypes.dataToPropertyValue(value).dataToEntry()
        val valueIdx = openCursor(txn)
        return valueIdx == null || SingleKeyCursorIsEmptyChecker(valueIdx, key).isEmpty
    }

    private inner class PropertyValueIterator(
        cursor: Cursor,
        val reverse: Boolean = false
    ) : EntityIteratorBase(this@PropertyValueIterable) {

        private var hasNext = false
        private val valueBytes: ArrayByteIterable

        init {
            setCursor(cursor)
            val binding = storeImpl.propertyTypes.dataToPropertyValue(value).binding
            valueBytes = binding.objectToEntry(value)
            val entry = cursor.getSearchKey(valueBytes)
            if (entry != null) {
                if (reverse) {
                    if (cursor.nextNoDup) {
                        checkHasNext(cursor.prev)
                    } else {
                        checkHasNext(cursor.last)
                    }
                } else {
                    checkHasNext(true)
                }
            }
        }

        public override fun hasNextImpl() = hasNext

        public override fun nextIdImpl(): EntityId? {
            if (hasNextImpl()) {
                explain(EntityIterableType.ENTITIES_BY_PROP_VALUE)
                val cursor = cursor
                return PersistentEntityId(entityTypeId, LongBinding.compressedEntryToLong(cursor.value)).also {
                    checkHasNext(if (reverse) cursor.prevNoDup else cursor.nextDup)
                }
            }
            return null
        }

        private fun checkHasNext(success: Boolean) {
            hasNext = success && valueBytes.compareTo(cursor.key) == 0
        }
    }
}
